File: mev-simulation/localnet/src/simulation/orchestrator.ts
import { Connection, Keypair, PublicKey } from "@solana/web3.js";
import chalk from "chalk";
import { SimulationConfig, lamportsToSol, formatSol } from "../config.js";
import { 
  SimulationResults, 
  SimulationSummary, 
  TradeResult, 
  SandwichResult,
  PoolStateRecord,
  PoolState 
} from "../types.js";
import { NormalTrader, randomTradeAmount, randomDirection } from "../bots/normal-trader.js";
import { ProtectedTrader } from "../bots/protected-trader.js";
import { SandwichAttacker } from "../bots/sandwich-attacker.js";
import { AccountSetupWithKeypair } from "../setup/accounts.js";
import { logSection, logProgress, endProgress, logOk, logger, log } from "../utils/logger.js";

/**
 * Simulation Orchestrator - Coordinates all simulation actors
 */
export class Orchestrator {
  private connection: Connection;
  private config: SimulationConfig;
  private poolAddress: PublicKey;
  private tokenAMint: PublicKey;
  private tokenBMint: PublicKey;

  private normalTraders: NormalTrader[] = [];
  private protectedTraders: ProtectedTrader[] = [];
  private attacker: SandwichAttacker | null = null;

  constructor(
    connection: Connection,
    config: SimulationConfig,
    poolAddress: PublicKey,
    tokenAMint: PublicKey,
    tokenBMint: PublicKey
  ) {
    this.connection = connection;
    this.config = config;
    this.poolAddress = poolAddress;
    this.tokenAMint = tokenAMint;
    this.tokenBMint = tokenBMint;
  }

  /**
   * Initialize traders and attacker
   */
  initialize(
    normalAccounts: AccountSetupWithKeypair[],
    protectedAccounts: AccountSetupWithKeypair[],
    attackerAccount: AccountSetupWithKeypair
  ): void {
    // Create normal traders
    for (const account of normalAccounts) {
      this.normalTraders.push(
        new NormalTrader(
          this.connection,
          account,
          this.poolAddress,
          this.tokenAMint,
          this.tokenBMint
        )
      );
    }

    // Create protected traders
    for (const account of protectedAccounts) {
      this.protectedTraders.push(
        new ProtectedTrader(
          this.connection,
          account,
          this.poolAddress,
          this.tokenAMint,
          this.tokenBMint
        )
      );
    }

    // Create attacker
    this.attacker = new SandwichAttacker(
      this.connection,
      attackerAccount,
      this.poolAddress,
      this.tokenAMint,
      this.tokenBMint
    );

    logOk("Orchestrator initialized");
    logger.info(`  Normal traders: ${this.normalTraders.length}`);
    logger.info(`  Protected traders: ${this.protectedTraders.length}`);
    logger.info(`  Attacker: ready`);
  }

  /**
   * Run the complete simulation
   */
  async run(): Promise<SimulationResults> {
    logSection("Running MEV Simulation");

    logger.info(`Comparison scenarios: ${this.config.transactions} (each scenario tests 2 trades)`);
    logger.info(`Total trades monitored: ${this.config.transactions * 2} (${this.config.transactions} normal + ${this.config.transactions} protected)`);
    logger.info(`Attack probability: ${this.config.attackProbability * 100}%`);
    logger.info(`Swap range: ${formatSol(this.config.minSwapLamports)} - ${formatSol(this.config.maxSwapLamports)} SOL`);
    log("");
    log(chalk.gray("  Each scenario compares: ") + chalk.yellow("Normal (vulnerable)") + " vs " + chalk.green("Protected (commit-reveal)"));
    log("");

    const normalTrades: TradeResult[] = [];
    const protectedTrades: TradeResult[] = [];
    const sandwichResults: SandwichResult[] = [];
    const poolHistory: PoolStateRecord[] = [];

    const startTime = Date.now();
    let totalOnChainTxs = 0;

    for (let i = 0; i < this.config.transactions; i++) {
      // Generate random trade parameters
      const amount = randomTradeAmount(
        this.config.minSwapLamports,
        this.config.maxSwapLamports
      );
      const direction = randomDirection();
      const shouldAttack = Math.random() < this.config.attackProbability;
      const directionStr = direction === "AtoB" ? "A→B" : "B→A";
      
      // Select random traders
      const normalTraderIdx = Math.floor(Math.random() * this.normalTraders.length);
      const protectedTraderIdx = Math.floor(Math.random() * this.protectedTraders.length);
      const normalTrader = this.normalTraders[normalTraderIdx];
      const protectedTrader = this.protectedTraders[protectedTraderIdx];

      // Get pool state before trades
      const poolBefore = await normalTrader.getPoolState();

      // ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
      log(chalk.cyan.bold(`\n━━━ Comparison Scenario ${i + 1}/${this.config.transactions} ━━━`));
      log(chalk.gray(`  Testing: ${formatSol(amount)} SOL ${directionStr} (same parameters for both approaches)\n`));

      // ═══════════════════════════════════════════════════════════
      // TRADE 1: Normal Trading (Vulnerable to MEV)
      // ═══════════════════════════════════════════════════════════
      log(chalk.yellow(`  [Trade ${i * 2 + 1}/${this.config.transactions * 2}] Normal Trade (Vulnerable):`));
      let normalResult: TradeResult;
      let sandwichResult: SandwichResult | null = null;
      let wasAttacked = false;

      if (shouldAttack && this.attacker) {
        // Simulate attacker seeing the pending swap in mempool
        const pendingSwap = {
          trader: normalTrader.publicKey,
          amountIn: amount,
          minOut: 0n,
          direction,
        };

        // === REALISTIC SANDWICH ATTACK SEQUENCE ===
        // Step 1: Attacker front-runs (swaps in same direction as victim)
        const frontRunResult = await this.attacker.executeFrontRun(pendingSwap, poolBefore);
        
        if (frontRunResult.success && frontRunResult.params) {
          wasAttacked = true;
          totalOnChainTxs++;
          log(chalk.red(`    ├─ [TX] 🔴 Front-run: Attacker swaps ${formatSol(frontRunResult.params.frontRunAmount)} SOL ${directionStr}`));
          
          // Step 2: Victim's trade executes (at worse price due to front-run)
          normalResult = await normalTrader.swap(amount, direction);
          totalOnChainTxs++;
          log(chalk.yellow(`    ├─ [TX] 🟡 Victim trade: ${formatSol(amount)} SOL ${directionStr} (worse price!)`));
          
          // Step 3: Attacker back-runs (swaps back to profit)
          sandwichResult = await this.attacker.executeBackRun(
            frontRunResult.params,
            frontRunResult.aToB,
            poolBefore
          );
          totalOnChainTxs++;
          
          if (sandwichResult.success) {
            sandwichResults.push(sandwichResult);
            log(chalk.red(`    ├─ [TX] 🔴 Back-run: Attacker swaps back, profit: ${formatSol(sandwichResult.profitLamports)} SOL`));
          }
          
          normalResult.wasAttacked = sandwichResult.success;
          if (sandwichResult.success) {
            normalResult.slippageLoss = sandwichResult.victimLossLamports;
          }
        } else {
          // Attack not profitable, victim trades normally
          normalResult = await normalTrader.swap(amount, direction);
          totalOnChainTxs++;
          log(chalk.gray(`    ├─ [TX] ⚪ Normal swap: ${formatSol(amount)} SOL ${directionStr} (attack not profitable)`));
          sandwichResult = {
            success: false,
            frontRunAmount: 0n,
            backRunAmount: 0n,
            profitLamports: 0n,
            victimLossLamports: 0n,
            reason: "Attack not profitable",
          };
        }
      } else {
        // No attack - normal swap
        normalResult = await normalTrader.swap(amount, direction);
        totalOnChainTxs++;
        log(chalk.gray(`    ├─ [TX] ⚪ Normal swap: ${formatSol(amount)} SOL ${directionStr}`));
      }
      normalTrades.push(normalResult);

      // Record pool state after normal scenario
      const poolAfterNormal = await normalTrader.getPoolState();
      poolHistory.push({
        transactionId: i,
        reserveA: poolAfterNormal.reserveA,
        reserveB: poolAfterNormal.reserveB,
        priceAInB: Number(poolAfterNormal.reserveB) / Number(poolAfterNormal.reserveA),
        scenario: "normal",
      });

      // Show result of normal trade
      if (wasAttacked && sandwichResult?.success) {
        log(chalk.red(`    └─ Result: ❌ ATTACKED - Lost ${formatSol(sandwichResult.victimLossLamports)} SOL\n`));
      } else {
        log(chalk.gray(`    └─ Result: ✓ No attack (not profitable)\n`));
      }

      // ═══════════════════════════════════════════════════════════
      // TRADE 2: Protected Trading (Commit-Reveal)
      // ═══════════════════════════════════════════════════════════
      log(chalk.green(`  [Trade ${i * 2 + 2}/${this.config.transactions * 2}] Protected Trade (Commit-Reveal):`));
      log(chalk.green(`    ├─ [TX] 🟢 Commit: Hash submitted (params hidden from MEV)`));
      totalOnChainTxs++;
      
      const protectedResult = await protectedTrader.commitAndReveal(amount, direction);
      protectedTrades.push(protectedResult);
      totalOnChainTxs += 2; // cancel + swap
      
      log(chalk.green(`    ├─ [TX] 🟢 Reveal+Swap: ${formatSol(amount)} SOL ${directionStr} (protected!)`));
      
      // Show savings from protection
      if (wasAttacked && sandwichResult?.success) {
        log(chalk.green.bold(`    └─ Result: ✓ PROTECTED - Saved ${formatSol(sandwichResult.victimLossLamports)} SOL from MEV\n`));
      } else {
        log(chalk.green(`    └─ Result: ✓ PROTECTED - No attack attempted\n`));
      }

      // Record pool state after protected scenario
      const poolAfterProtected = await protectedTrader.getPoolState();
      poolHistory.push({
        transactionId: i,
        reserveA: poolAfterProtected.reserveA,
        reserveB: poolAfterProtected.reserveB,
        priceAInB: Number(poolAfterProtected.reserveB) / Number(poolAfterProtected.reserveA),
        scenario: "protected",
      });

      // Scenario summary with timing
      const elapsed = (Date.now() - startTime) / 1000;
      const rate = (i + 1) / elapsed;
      log(chalk.gray(`  ⏱️  Completed scenario ${i + 1}/${this.config.transactions} (${rate.toFixed(1)} scenarios/s)\n`));
    }

    log(chalk.cyan(`  Total on-chain transactions: ${totalOnChainTxs}`));

    // Calculate summary statistics
    const summary = this.calculateSummary(normalTrades, protectedTrades, sandwichResults);

    const totalTime = (Date.now() - startTime) / 1000;
    logOk(`Simulation complete in ${totalTime.toFixed(1)}s`);

    return {
      config: {
        transactions: this.config.transactions,
        attackProbability: this.config.attackProbability,
        minSwapLamports: this.config.minSwapLamports.toString(),
        maxSwapLamports: this.config.maxSwapLamports.toString(),
        initialPoolLiquidity: this.config.initialPoolLiquidity.toString(),
        feeBps: this.config.feeBps,
      },
      normalTrades,
      protectedTrades,
      sandwichResults,
      summary,
      poolHistory,
    };
  }

  /**
   * Calculate summary statistics
   */
  private calculateSummary(
    normalTrades: TradeResult[],
    protectedTrades: TradeResult[],
    sandwichResults: SandwichResult[]
  ): SimulationSummary {
    const totalTransactions = normalTrades.length;
    const attackAttempts = sandwichResults.length;
    const successfulAttacks = sandwichResults.filter(s => s.success).length;
    const attackSuccessRate = attackAttempts > 0 
      ? (successfulAttacks / attackAttempts) * 100 
      : 0;

    const normalTransactions = normalTrades.length;
    const normalAttacked = normalTrades.filter(t => t.wasAttacked).length;
    const protectedTransactions = protectedTrades.length;
    const protectedAttacked = 0; // Commit-reveal is not attackable

    const totalMevExtracted = sandwichResults
      .filter(s => s.success)
      .reduce((sum, s) => sum + s.profitLamports, 0n);

    const totalVictimLosses = sandwichResults
      .filter(s => s.success)
      .reduce((sum, s) => sum + s.victimLossLamports, 0n);

    const avgLossPerAttack = successfulAttacks > 0
      ? lamportsToSol(totalVictimLosses) / successfulAttacks
      : 0;

    // Protected savings = what victims would have lost without protection
    const totalProtectedSavings = totalVictimLosses;

    const totalVolume = normalTrades.reduce((sum, t) => sum + t.amountIn, 0n);
    const avgTradeAmount = totalTransactions > 0
      ? lamportsToSol(totalVolume) / totalTransactions
      : 0;

    return {
      totalTransactions,
      attackAttempts,
      successfulAttacks,
      attackSuccessRate,
      normalTransactions,
      normalAttacked,
      protectedTransactions,
      protectedAttacked,
      totalMevExtracted,
      totalVictimLosses,
      avgLossPerAttack,
      totalProtectedSavings,
      avgTradeAmount,
      totalVolume,
    };
  }
}

/**
 * Print simulation summary to console
 */
export function printSummary(results: SimulationResults): void {
  const s = results.summary;

  console.log(chalk.blue("\n╔══════════════════════════════════════════════════════════════════╗"));
  console.log(chalk.blue("║            MEV SIMULATION RESULTS                                ║"));
  console.log(chalk.blue("╠══════════════════════════════════════════════════════════════════╣"));
  
  console.log(chalk.blue("║                                                                  ║"));
  console.log(chalk.blue("║  CONFIGURATION                                                   ║"));
  console.log(chalk.gray(`║  Comparison Scenarios:   ${s.totalTransactions.toString().padEnd(10)} (each tests 2 approaches)  ║`));
  console.log(chalk.gray(`║  Normal Trades Tested:   ${s.normalTransactions.toString().padEnd(10)}                          ║`));
  console.log(chalk.gray(`║  Protected Trades Tested:${s.protectedTransactions.toString().padEnd(10)}                          ║`));
  console.log(chalk.gray(`║  Total Trades Monitored: ${(s.normalTransactions + s.protectedTransactions).toString().padEnd(10)}                          ║`));
  console.log(chalk.gray(`║  Attack Probability:     ${(results.config.attackProbability * 100).toFixed(0)}%                               ║`));
  console.log(chalk.gray(`║  Pool Fee:               ${(results.config.feeBps / 100).toFixed(2)}%                              ║`));

  console.log(chalk.blue("║                                                                  ║"));
  console.log(chalk.blue("╠══════════════════════════════════════════════════════════════════╣"));
  console.log(chalk.blue("║                                                                  ║"));
  console.log(chalk.blue("║  NORMAL TRADING (Vulnerable to MEV)                              ║"));
  console.log(chalk.red(`║  Attack Attempts:        ${s.attackAttempts.toString().padEnd(10)}                          ║`));
  console.log(chalk.red(`║  Successful Attacks:     ${s.successfulAttacks.toString().padEnd(10)}                          ║`));
  console.log(chalk.red(`║  Attack Success Rate:    ${s.attackSuccessRate.toFixed(1)}%                              ║`));
  console.log(chalk.red(`║  Trades Attacked:        ${s.normalAttacked}/${s.normalTransactions} normal trades attacked           ║`));
  console.log(chalk.red(`║  Total MEV Extracted:    ${formatSol(s.totalMevExtracted).padEnd(10)} SOL                    ║`));
  console.log(chalk.red(`║  Total Victim Losses:    ${formatSol(s.totalVictimLosses).padEnd(10)} SOL                    ║`));
  console.log(chalk.red(`║  Avg Loss per Attack:    ${s.avgLossPerAttack.toFixed(6).padEnd(10)} SOL                    ║`));

  console.log(chalk.blue("║                                                                  ║"));
  console.log(chalk.blue("╠══════════════════════════════════════════════════════════════════╣"));
  console.log(chalk.blue("║                                                                  ║"));
  console.log(chalk.blue("║  PROTECTED TRADING (Commit-Reveal)                               ║"));
  console.log(chalk.green(`║  Trades Attacked:        ${s.protectedAttacked}/${s.protectedTransactions} protected trades attacked      ║`));
  console.log(chalk.green(`║  Attacks Possible:       0                                        ║`));
  console.log(chalk.green(`║  MEV Extracted:          0 SOL                                    ║`));
  console.log(chalk.green(`║                                                                  ║`));
  console.log(chalk.green(`║  ★ TOTAL SAVINGS:        ${formatSol(s.totalProtectedSavings).padEnd(10)} SOL                    ║`));
  console.log(chalk.green(`║  ★ Protection Rate:      100.0%                                   ║`));

  console.log(chalk.blue("║                                                                  ║"));
  console.log(chalk.blue("╠══════════════════════════════════════════════════════════════════╣"));
  console.log(chalk.blue("║                                                                  ║"));
  console.log(chalk.blue("║  VOLUME STATISTICS                                               ║"));
  console.log(chalk.gray(`║  Total Volume:           ${formatSol(s.totalVolume).padEnd(10)} SOL                    ║`));
  console.log(chalk.gray(`║  Average Trade:          ${s.avgTradeAmount.toFixed(4).padEnd(10)} SOL                    ║`));
  console.log(chalk.blue("║                                                                  ║"));
  console.log(chalk.blue("╚══════════════════════════════════════════════════════════════════╝"));
  console.log();
}

File: mev-simulation/src/simulation/orchestrator.rs
//! Simulation Orchestrator
//!
//! Controls the flow of the MEV simulation, running both scenarios
//! (vulnerable normal trading vs protected commit-reveal) and collecting results.

use crate::bots::{
    normal_trader::{NormalTrader, TradeResult, random_trade_amount, random_direction},
    protected_trader::ProtectedTrader,
    sandwich_attacker::{SandwichAttacker, SandwichResult, PendingSwap},
};
use crate::config::SimulationConfig;
use crate::simulation::pool_state::SimulatedPool;
use anyhow::Result;
use rand::Rng;
use serde::{Deserialize, Serialize};
use solana_sdk::signature::Keypair;
#[allow(unused_imports)]
use solana_sdk::signer::Signer;
use tracing::{info, debug};

/// Results of the complete simulation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimulationResults {
    /// Configuration used
    pub config: SimulationConfigSummary,
    /// Results from normal (vulnerable) trades
    pub normal_trades: Vec<TradeResult>,
    /// Results from protected trades
    pub protected_trades: Vec<TradeResult>,
    /// Sandwich attack results
    pub sandwich_results: Vec<SandwichResult>,
    /// Summary statistics
    pub summary: SimulationSummary,
    /// Pool state history
    pub pool_history: Vec<PoolStateRecord>,
}

/// Summary of simulation configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimulationConfigSummary {
    pub total_transactions: u32,
    pub attack_probability: f64,
    pub min_swap_lamports: u64,
    pub max_swap_lamports: u64,
    pub initial_pool_a: u64,
    pub initial_pool_b: u64,
    pub fee_bps: u16,
}

/// Summary statistics
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct SimulationSummary {
    /// Total transactions simulated
    pub total_transactions: u32,
    /// Number of attack attempts
    pub attack_attempts: u32,
    /// Number of successful attacks
    pub successful_attacks: u32,
    /// Attack success rate (%)
    pub attack_success_rate: f64,
    /// Total MEV extracted (lamports)
    pub total_mev_extracted: i64,
    /// Total victim losses (lamports)
    pub total_victim_losses: u64,
    /// Average loss per attacked transaction (lamports)
    pub avg_loss_per_attack: f64,
    /// Total protected savings (lamports) - what users saved by using commit-reveal
    pub total_protected_savings: u64,
    /// Average trade amount (lamports)
    pub avg_trade_amount: f64,
    /// Total volume traded (lamports)
    pub total_volume: u64,
}

/// Record of pool state at a point in time
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PoolStateRecord {
    pub transaction_id: u32,
    pub reserve_a: u64,
    pub reserve_b: u64,
    pub price_a_in_b: f64,
    pub scenario: String,
}

/// Main simulation orchestrator
pub struct Orchestrator {
    /// Configuration
    config: SimulationConfig,
    /// Sandwich attacker bot
    attacker: SandwichAttacker,
    /// Normal trader (for vulnerable scenario)
    normal_traders: Vec<NormalTrader>,
    /// Protected traders (for commit-reveal scenario)
    protected_traders: Vec<ProtectedTrader>,
    /// Simulated pool
    pool: SimulatedPool,
    /// Current transaction counter
    transaction_counter: u32,
}

impl Orchestrator {
    /// Create a new orchestrator with the given configuration
    pub fn new(config: SimulationConfig) -> Self {
        // Create attacker
        let attacker_keypair = Keypair::new();
        let attacker = SandwichAttacker::new(
            attacker_keypair,
            config.attacker_capital,
            config.attacker_capital, // Give attacker both tokens
        );

        // Create normal traders
        let mut normal_traders = Vec::new();
        for _ in 0..config.num_victims {
            let keypair = Keypair::new();
            let trader = NormalTrader::new(
                keypair,
                config.victim_sol,
                config.victim_secusol,
            );
            normal_traders.push(trader);
        }

        // Create protected traders (same balances as normal traders)
        let mut protected_traders = Vec::new();
        for _ in 0..config.num_victims {
            let keypair = Keypair::new();
            let trader = ProtectedTrader::new(
                keypair,
                config.victim_sol,
                config.victim_secusol,
            );
            protected_traders.push(trader);
        }

        // Create pool
        let pool = SimulatedPool::new(
            config.initial_pool_a,
            config.initial_pool_b,
            config.fee_bps,
        );

        Self {
            config,
            attacker,
            normal_traders,
            protected_traders,
            pool,
            transaction_counter: 0,
        }
    }

    /// Run the complete simulation
    pub fn run(&mut self) -> Result<SimulationResults> {
        info!("Starting MEV simulation...");
        info!("Total transactions: {}", self.config.total_transactions);
        info!("Attack probability: {:.0}%", self.config.attack_probability * 100.0);

        let mut normal_trades = Vec::new();
        let mut protected_trades = Vec::new();
        let mut sandwich_results = Vec::new();
        let mut pool_history = Vec::new();

        let mut rng = rand::thread_rng();

        for i in 0..self.config.total_transactions {
            self.transaction_counter = i;

            // Generate random trade parameters
            let amount = random_trade_amount(
                self.config.min_swap_lamports,
                self.config.max_swap_lamports,
            );
            let a_to_b = random_direction();
            let trader_idx = rng.gen_range(0..self.normal_traders.len());

            // Decide if attacker will attempt a sandwich
            let should_attack = rng.gen::<f64>() < self.config.attack_probability;

            // === SCENARIO A: Normal Trading (Vulnerable) ===
            // Save pool state before normal scenario
            let pool_state_before = self.pool.clone_state();

            let normal_trade = self.run_normal_scenario(
                trader_idx,
                amount,
                a_to_b,
                should_attack,
                &mut sandwich_results,
            );

            if let Some(trade) = normal_trade {
                normal_trades.push(trade);
            }

            // Record pool state after normal scenario
            pool_history.push(PoolStateRecord {
                transaction_id: i,
                reserve_a: self.pool.current().reserve_a,
                reserve_b: self.pool.current().reserve_b,
                price_a_in_b: self.pool.price_a_in_b(),
                scenario: "normal".to_string(),
            });

            // === SCENARIO B: Protected Trading (Commit-Reveal) ===
            // Reset pool to same state as before normal scenario
            self.pool.set_state(pool_state_before);

            let protected_trade = self.run_protected_scenario(
                trader_idx,
                amount,
                a_to_b,
            );

            if let Some(trade) = protected_trade {
                protected_trades.push(trade);
            }

            // Record pool state after protected scenario
            pool_history.push(PoolStateRecord {
                transaction_id: i,
                reserve_a: self.pool.current().reserve_a,
                reserve_b: self.pool.current().reserve_b,
                price_a_in_b: self.pool.price_a_in_b(),
                scenario: "protected".to_string(),
            });

            // Progress logging
            if (i + 1) % 100 == 0 || i == 0 {
                info!("Progress: {}/{} transactions", i + 1, self.config.total_transactions);
            }
        }

        // Calculate summary statistics
        let summary = self.calculate_summary(&normal_trades, &protected_trades, &sandwich_results);

        info!("Simulation complete!");
        info!("Total MEV extracted: {} lamports ({:.4} SOL)", 
              summary.total_mev_extracted,
              summary.total_mev_extracted as f64 / 1_000_000_000.0);
        info!("Total victim losses: {} lamports ({:.4} SOL)",
              summary.total_victim_losses,
              summary.total_victim_losses as f64 / 1_000_000_000.0);

        Ok(SimulationResults {
            config: SimulationConfigSummary {
                total_transactions: self.config.total_transactions,
                attack_probability: self.config.attack_probability,
                min_swap_lamports: self.config.min_swap_lamports,
                max_swap_lamports: self.config.max_swap_lamports,
                initial_pool_a: self.config.initial_pool_a,
                initial_pool_b: self.config.initial_pool_b,
                fee_bps: self.config.fee_bps,
            },
            normal_trades,
            protected_trades,
            sandwich_results,
            summary,
            pool_history,
        })
    }

    /// Run the normal (vulnerable) trading scenario
    fn run_normal_scenario(
        &mut self,
        trader_idx: usize,
        amount: u64,
        a_to_b: bool,
        should_attack: bool,
        sandwich_results: &mut Vec<SandwichResult>,
    ) -> Option<TradeResult> {
        let trader = &mut self.normal_traders[trader_idx];

        // Check if trader has sufficient balance
        if !trader.can_trade(amount, a_to_b) {
            debug!("Trader {} has insufficient balance", trader_idx);
            return None;
        }

        // Calculate expected output BEFORE any attack
        let expected_out = trader.calculate_expected(amount, a_to_b, self.pool.current());

        let was_attacked = if should_attack {
            // Create pending swap that attacker can see
            let pending = PendingSwap {
                amount_in: amount,
                a_to_b,
                victim: trader.pubkey(),
                min_out: 0, // Normal trades often don't set this properly
            };

            // Attacker executes sandwich
            let sandwich = self.attacker.execute_sandwich(&pending, self.pool.current_mut());
            
            let attacked = sandwich.success;
            sandwich_results.push(sandwich);
            attacked
        } else {
            // No attack, just execute victim's trade
            let _ = self.pool.current_mut().apply_swap(amount, a_to_b);
            false
        };

        // Now the victim's trade has already been executed as part of the sandwich
        // We just need to record the result
        Some(TradeResult {
            signature: format!("normal_trade_{}", self.transaction_counter),
            trader: trader.pubkey().to_string(),
            amount_in: amount,
            a_to_b,
            expected_out,
            actual_out: if was_attacked {
                // The actual output was calculated during sandwich
                let last_sandwich = sandwich_results.last().unwrap();
                expected_out.saturating_sub(last_sandwich.victim_loss_lamports)
            } else {
                expected_out
            },
            slippage_loss: if was_attacked {
                sandwich_results.last().unwrap().victim_loss_lamports
            } else {
                0
            },
            was_attacked,
            fee_paid: (amount as u128 * self.config.fee_bps as u128 / 10000) as u64,
            price_impact_bps: 0,
            timestamp: chrono::Utc::now().timestamp(),
        })
    }

    /// Run the protected (commit-reveal) trading scenario
    fn run_protected_scenario(
        &mut self,
        trader_idx: usize,
        amount: u64,
        a_to_b: bool,
    ) -> Option<TradeResult> {
        let trader = &mut self.protected_traders[trader_idx];

        // Execute protected trade
        let result = trader.execute_protected_trade(
            amount,
            a_to_b,
            self.pool.current_mut(),
            100, // 1% slippage tolerance
        )?;

        // Return the underlying trade result
        Some(result.trade)
    }

    /// Calculate summary statistics
    fn calculate_summary(
        &self,
        normal_trades: &[TradeResult],
        _protected_trades: &[TradeResult],
        sandwich_results: &[SandwichResult],
    ) -> SimulationSummary {
        let total_transactions = normal_trades.len() as u32;
        let attack_attempts = sandwich_results.len() as u32;
        let successful_attacks = sandwich_results.iter()
            .filter(|s| s.success)
            .count() as u32;

        let attack_success_rate = if attack_attempts > 0 {
            (successful_attacks as f64 / attack_attempts as f64) * 100.0
        } else {
            0.0
        };

        let total_mev_extracted: i64 = sandwich_results.iter()
            .map(|s| s.profit_lamports)
            .sum();

        let total_victim_losses: u64 = sandwich_results.iter()
            .map(|s| s.victim_loss_lamports)
            .sum();

        let avg_loss_per_attack = if successful_attacks > 0 {
            total_victim_losses as f64 / successful_attacks as f64
        } else {
            0.0
        };

        // Protected savings = what victims would have lost without protection
        let total_protected_savings = total_victim_losses;

        let total_volume: u64 = normal_trades.iter()
            .map(|t| t.amount_in)
            .sum();

        let avg_trade_amount = if total_transactions > 0 {
            total_volume as f64 / total_transactions as f64
        } else {
            0.0
        };

        SimulationSummary {
            total_transactions,
            attack_attempts,
            successful_attacks,
            attack_success_rate,
            total_mev_extracted,
            total_victim_losses,
            avg_loss_per_attack,
            total_protected_savings,
            avg_trade_amount,
            total_volume,
        }
    }

    /// Reset the orchestrator for another run
    pub fn reset(&mut self) {
        self.attacker.reset(
            self.config.attacker_capital,
            self.config.attacker_capital,
        );

        for trader in &mut self.normal_traders {
            trader.reset(self.config.victim_sol, self.config.victim_secusol);
        }

        for trader in &mut self.protected_traders {
            trader.reset(self.config.victim_sol, self.config.victim_secusol);
        }

        self.pool.reset();
        self.transaction_counter = 0;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_orchestrator_quick_run() {
        let config = SimulationConfig {
            total_transactions: 10,
            ..SimulationConfig::quick_test()
        };

        let mut orchestrator = Orchestrator::new(config);
        let results = orchestrator.run().unwrap();

        assert_eq!(results.normal_trades.len(), 10);
        assert_eq!(results.protected_trades.len(), 10);
        
        println!("MEV extracted: {} lamports", results.summary.total_mev_extracted);
        println!("Victim losses: {} lamports", results.summary.total_victim_losses);
        println!("Attack success rate: {:.1}%", results.summary.attack_success_rate);
    }
}

